source/module_basis/module_pw/test_serial/pw_basis_k_test.cpp

```diff
@@ -190,3 +190,25 @@ TEST_F(PWBasisKTEST, CollectLocalPW)
 
 #undef private
 #undef protected
+
+// It is weird to init MPI again in this serial test, but it is necessary
+// to avoid "Attempting to use an MPI routine before initializing MPICH" in intel compiler.
```

Comment:
Would calling `mpi_initialized` ahead of `MPI_finalize` helps?

Reply:
Hi @caic99, thanks for your suggestion. After consulting chat expert, I get to know that mpi_initialized is used to check if MPI has been initialized. But in this test, the error happens because MPI is used before MPI is initialized. So I think that call mai_initiazlied will not help. What's your opinion?

Reply:
> Hi @caic99, thanks for your suggestion. After consulting chat expert, I get to know that mpi_initialized is used to check if MPI has been initialized. But in this test, the error happens because MPI is used before MPI is initialized. So I think that calling mai_initiazlied will not help. What's your opinion?

@hongriTianqi We can call `mpi_initialized` before calling the erroneous MPI routine to ensure it is initialized, otherwise skipping. Would that be better?

Reply:
Hi @caic99, thanks. However, after try as you said, the unit test cannot pass. What I have did is adding the `mpi_initialized` in SetUp:
```
void SetUp()
          {
                int is_initialized;
                MPI_Initialized(&is_initialized);
          }
```
So, do you have better solutions, please?

Reply:
> Hi @caic99, thanks. However, after trying as you said, the unit test cannot pass. What I have did is adding the `mpi_initialized` in SetUp:
> 
> ```
> void SetUp()
> {
>       int is_initialized;
>       MPI_Initialized(&is_initialized);
> }
> ```
> 
> So, do you have better solutions, please?

If that MPI operations are the same one, you can try modifying them like below:
```c++
int is_initialized;
MPI_Initialized(&is_initialized);
if(is_initialized)
{
    MPI_finalize();
}
```

However, if the errors are everywhere and impossible to wrap them all, I think your solution would be fine.